src/sentry/audit_log/manager.py

```diff
@@ -1,49 +1,111 @@
 from dataclasses import dataclass
-from typing import Callable, List
+from typing import List, Optional
 
-import sentry_sdk
+from sentry_sdk import capture_exception
 
 from sentry.models.auditlogentry import AuditLogEntry
 
 
-@dataclass
+class DuplicateAuditLogEvent(Exception):
+    pass
+
+
+class AuditLogEventNotRegistered(Exception):
+    pass
+
+
+"""
+The audit log system records changes made to an organization and displays them in
+organization settings.
+
+To add a new audit log event:
+
+1. Create a new instance of AuditLogEvent. You'll need an event_id, name, api_name,
+   and optional template.
+
+   Note: The template uses AuditLogEntry.data fields to construct a simple audit
+   log message. For more complicated messages, subclass AuditLogEvent in events.py
+   and override the render function.
+
+2. Register the AuditLogEvent using `default_manager.add()`.
+
+    default_manager.add(
+        AuditLogEvent(
+            event_id=1,
+            name="MEMBER_INVITE",
+            api_name="member.invite",
+            template="invited member {email}",
+        )
+    )
+
+3. Record your AuditLogEvent.
+
+    self.create_audit_entry(
+        request=request,
+        organization_id=organization.id,
+        target_object=member.id,
+        data={"email": "email@gmail.com"},
+        event=audit_log.get_event_id(MEMBER_INVITE),
+    )
+"""
+
+
+@dataclass(init=False)
 class AuditLogEvent:
+    # Unique event ID (ex. 1)
     event_id: int
+
+    # Unique event name (ex. MEMBER_INVITE)
     name: str
+
+    # Unique event api name (ex. member.invite)
     api_name: str
-    # render holds a function that will determine the message to be displayed in the audit log
-    render: Callable[[AuditLogEntry], str]
+
+    # Simple template for rendering the audit log message using
+    # the AuditLogEntry.data fields. For more complicated messages,
+    # subclass AuditLogEvent and override the render method.
+    template: Optional[str] = None
+
+    def __init__(self, event_id, name, api_name, template=None):
+        self.event_id = event_id
+        self.name = name
+        self.api_name = api_name
+        self.template = template
+
+    def render(self, audit_log_entry: AuditLogEntry):
```

Comment:
:+1: I like this approach better. It makes both simple audits and complex ones declarative, as simple events are defined inline and more complex ones are get classes.